File: LCA/src/LCA.java
public class LCA {
    // Defining a node in a binary tree.
    public class Node {
        Node left;
        Node right;
    }

    public Node lowestCommonAncestor(Node node1, Node node2, Node root)
    {
        // If one of the nodes has been found (i.e. is root), return root
        if (node1 == root || node2 == root) return root;

        // Traverse tree from left and right to look for nodes
        Node leftLCA = lowestCommonAncestor(node1, node2, root.left);
        Node rightLCA = lowestCommonAncestor(node1, node2, root.right);

        // If node1 and node2 are both found in subtrees of root, then root is the LCA.
        if (leftLCA != null && rightLCA != null) return root;

        // Else... return either left or right if the other is not found.
        if (leftLCA == null)
            return rightLCA;
        else
            return leftLCA;
    }
}

File: LCA/src/LCA.py
# PYTHON 3.9x
# LOWEST COMMON ANCESTOR
# LIAM O LIONAIRD [19335530]
# v1: 25/09/21
# v2: 14/10/21

# written with Vim, submitted using git command-line!

import unittest

class Node:
    def __init__(self):
        self.left = None # when testing, set these to a Node object
        self.right = None

    def setLeft(self, left):
        self.left = left

    def setRight(self, right):
        self.right = right

    def getLeft(self):
        return self.left

    def getRight(self):
        return self.right

def lowestCommonAncestor(node1, node2, root):
    # adapting code from LCA.java...

    if root == None:
        return None
    if node1 == root or node2 == root:
        return root

    leftLCA = lowestCommonAncestor(node1, node2, root.getLeft())
    rightLCA = lowestCommonAncestor(node1, node2, root.getRight())

    if leftLCA != None and rightLCA != None:
        return root

    if leftLCA is None:
        return rightLCA
    else:
        return leftLCA

class testLCA(unittest.TestCase):

    def test_node(self):
        a = Node()
        b = Node()
        c = Node()
        a.setLeft(b)
        a.setRight(c)
        self.assertEqual(a.getLeft(), b)
        self.assertEqual(a.getRight(), c)

    def test_LCA(self):
        # Arranging these nodes into the following tree:
        #        A
        #       / \
        #      B   C
        #     / \   \
        #    D   E   F
        #       /
        #      G
        a = Node()
        b = Node()
        c = Node()
        d = Node()
        e = Node()
        f = Node()
        g = Node()
        a.setLeft(b)
        a.setRight(c)
        b.setLeft(d)
        b.setRight(e)
        c.setRight(f)
        e.setLeft(g)

        self.assertEqual(a, lowestCommonAncestor(a, a, a))
        self.assertEqual(a, lowestCommonAncestor(b, c, a))
        self.assertEqual(a, lowestCommonAncestor(d, f, a))
        self.assertEqual(b, lowestCommonAncestor(d, g, a))

if __name__ == '__main__':
    unittest.main() # Runs the tests
